Add initial defunct multi-package support for RFC

bschindler opened the pull request:
I wanted to create an issue with a question, but since issues seem to be disabled, I'd like to do this using pull requests. This is not meant for immediate merging as it is not completely finished (read, not working), but good enough to have the discussion. 

In order to facilitate switching between release and debug (and other build modes) on Visual Studio without rerunning conan, we (meaning the company I work for) create so called multi-packages. The package for windows will contain both release and debug version at the same time, thus build_type and runtime are removed from settings (see config_options method). In build, the package is just built 2 times and in the packaging, both libraries have to be added. For the runtime, the dynamic multithreaded debug and non-debug are used for debug and release respectively. This is not done for linux and other environments as a build type uses one single build folder. 

Since essentially no package supports this in conan-central, I thought about raising this topic because if accepted, it would prevent us from having to maintain a fork of every recipe we require for our internal software. And I think the ability to switch build modes without rerunning has also benefits outside of what we do... 

So, what do you think? If you are favorable to such a change, I can finish it up and get it into a mergeable state such that it could be included in this recipe.

bschindler committed:
Add initial defunct multi-package support for RFC

solvingj commented:
Hi @bschindler , I'm terribly sorry for the inconvenience of the disabled issues.  We have not yet managed to get back around to all our README files and update them with some new information. 

We've centralized all our issues in the repository below, so please use it for opening any issues on Bincrafters packages in the future. 

https://github.com/bincrafters/community/issues

We are currently working on a Visual Studio plugin that is currently awaiting testing.  I will ping the author to ask him to come comment on whether or not it might alleviate your problems.  

My initial feeling is that I understand why you are doing the multi_packages, but I think the tools exist to make that an unnecessary compromise (and it is a significant compromise).  From a distance (not having struggled with the problem myself), it feels like the way Conan currently organizes the binaries (fully separated and modularized) is the correct and optimal approach.  With this point of view, I would say the best way to approach your problem is to try to use the tooling to optimize your workflow "around" the Conan binaries and client.

bschindler commented:
Hi @solvingj do you have a pointer to this visual studio plugin? I'm certainly interested and would like to see the ideas

As for what you write about the multi_packages, I fully agree that the approach I propose is not the nicest. I had a lengthy discussion with the conan devs here: https://github.com/conan-io/conan/issues/917 and I'm as unhappy with the situation as you are I suppose. Yet, I am always puzzled when I hear people say that this approach is "correct/optimal/whatever", simply because there is no correct way of organizing binaries. In cmake, for some generators you specify the build type at configure time (which is kinda where conan is expected to run), but also generators where the build type is specified at build time (Visual Studio and some others as well). In an ideal world everybody would agree on where/how build types/configurations are handled. Unfortunately, reality strikes back :(

Specifying the build type at build time is simply how things work under Visual Studio. If the expectation is that for switching build types the user has to run conan apart from switching build type in the IDE (even with some kind of automation), the result is that cmake will rerun, changing include locations forcing a full rebuild of whatever you are working on. That is simply not an acceptable solution and will make windows developers walk away (as I've experienced in my company). In that sense, I don't really see much optimization potential from what I write above. 

I'm open to other suggestions - integrating conan into cmake at build time or making it possible to have packages which explicitly list their contents (release and debug, runtime xxx etc) is all a solution. What is not a solution is additional work for switching build types and rebuilds triggered due to executing conan/cmake

solvingj commented:
Here is the link: 
https://github.com/bincrafters/conan-vs-extension

Indeed, describing a strategy as "correct" is an inappropriate choice of words. However, I think it is "optimal" is appropriate, as the current strategy is likely to be the most advantageous overall (considering the many known factors in the many known workflows, IDE's, and usage patterns). 

I agree with you regarding what would be acceptable, and that a full rebuild upon switching build-type in visual studio is not.  Based on what I do know, I am confident that the switch can be nearly transparent and seamless within the current Conan features.   There is a caveat, in that there are multiple ways to use Visual Studio + Cmake + Conan together.  It's possible that one or two of these ways might not support such a transparent worfklow.  I don't know for sure, that's just a disclaimer since I haven't tried them all. 

If you don't mind, I'd like to move this discussion to a github issue on the community repo, and keep it open until I can get back to you with some concrete examples.

ghost commented:
@bschindler are you aware of conan's documented "N configs -> 1 package" functionality? It is described here:

http://docs.conan.io/en/latest/creating_packages/package_approaches.html?highlight=debug#n-configs-1-package

I have never tried using it but it seems to me that it is supposed to solve the problem you are describing. If you know about it and have discovered some deficiencies which have led you to develop your own method please let us know. Otherwise how about we try it out and potentially add it to our templates with @solvingj  and the rest of the team?

bschindler commented:
@solvingj Sure, lets move the discussion to an issue on the community page. Until this is open:

@dimi309: Isn't this exactly what I'm doing? Apart from setting the build_type to ANY (which didn't exist when I started with the multi-packages), that's what I'm using. This section in the documentation is a result of the issue I linked from conan afict. 
Yes, it does more or less solve the issue I'm describing. The "less" part stems from the fact that setting build type to ANY makes conan unaware of what is contained. Its just more or less a convention that Release and Debug are part of it and that the multithreaded dynamic runtime is used (debug for Debug and non-debug for release), so this method is a bit implicit. 

It would be absolutely fantastic if bincrafters would add some support to multi-packages. I'm sure my team would be willing to upstream our changes to your packages. It would mean we would not have to maintain forks of packages and you get multi-package support which is benefitial to VS users. 
Currently, we maintain forks of boost, glfw, gtest, google benchmark, Eigen, gsl, spdlog, foonathan_memory and some others which are most likely not on bincrafters

ghost commented:
@bschindler I considered the removal of build_type and runtime to be the discerning factor between your way and the one described in the documentation. I may be wrong. Whichever point of view is correct, before skipping over part of conan's suggested method, perhaps we should contact them first to see if we can reach a common understanding? It might take longer but it I think that it will serve package quality and user comprehension in the long run.

bschindler commented:
@dimi309 I just tried the multi example which is mentioned in the docs. Setting ANY in the package_id essentially removes the build_type from the conaninfo.txt. So the package is going to be the same, writing ANY is probably a nicer way of achieving this than what I'm proposing here. I'll ask conan about the runtime as this flag does not make sense in that sutation but is specified. 

For me the more interesting question is when/how multi-packages are created. The pull request here makes a multi-config only for IDE's having multi-config, the example on conan does it in any case (which I find a bit odd tbh). Should it always be enabled for VS or should it be an option to retain flexibility?

bschindler commented:
https://github.com/conan-io/conan/issues/2328
It seems runtime should be removed as well from the package_id - will supposedly be fixed for 1.1

ForNeVeR commented:
@bschindler hello, I'm one of the authors of the Visual Studio plugin mentioned earlier. I'd like to better understand your use case with Conan + Visual Studio, and maybe give you advice about the workflow that's proposed by the plugin _(the development version of the plugin is available [from my fork](https://github.com/ForNeVeR/conan-vs-extension/) but you'll need to build it yourself, no releases are provided yet)_.

Are you aware of [`visual_studio_multi`]( http://docs.conan.io/en/latest/reference/generators/visualstudiomulti.html) Conan generator? It looks to me that it _could_ offer a solution to your problem regarding multiple configurations.

Summarily, you could run `conan install ...` with multiple configurations, and they all will be preserved in `conanbuildinfo.props` file (that you should wire into your `vcxproj`). After that, you can switch the configurations in Visual Studio and build them, and they will just work (every Release/Debug/x64/whatever config will use the properly built artifacts from **separately built** version of Conan package).

My plugin automates some of the one-time manual work: calling `conan install` with proper settings for each of the project configurations and tuning the `vcxproj` file, but you could use `visual_studio_multi` even without the plugin. Does it help?

bschindler commented:
@ForNeVeR I am not using Visual Studio directly. I am working on a cmake based project where the majority of developers code on it using Visual Studio. So the visual_studio_multi generator does not apply. Sorry for not mentioning this earlier. 
What would theoreticaly apply is the cmake_multi generator (as mentioned in the linked conan issue above). This has however the issue that it breaks many standard FindXXX scripts (boost in particular) as they expect debug/release libraries to be located at the same place.

solvingj commented:
@bschindler please explain how your developers open the project from visual studio. Do they use the new Cmake support in Visual studio, or are they on an older version so they just browse to and open the `vcxproj` files directly, or do they use one of the older techniques of using Cmake and visual studio together?

ghost commented:
So, @bschindler, while we are discussing everything else, I am trying to create a "reference implementation" so to speak of what we are talking about, so that we can think about this based on something concrete. I have cloned the conan-glfw repository and incorporated your pull request here: https://github.com/dimi309/conan-glfw I am trying to get it working, first using your method exactly as you describe it. Then, we can explore a bit making it 100% according to the conan documentation in order to check for any differences or drawbacks in that case.

Right now I am getting some unresolved external symbol errors:

```
3>test_package.obj : error LNK2019: unresolved external symbol glfwTerminate referenced in function main [C:\Users
\user\Source\conan-glfw\test_package\build\99c165bdad86231845cc4567bbeaedfcde722ed4\test_package.vcxproj]
3>test_package.obj : error LNK2019: unresolved external symbol glfwGetPrimaryMonitor referenced in function main [
C:\Users\user\Source\conan-glfw\test_package\build\99c165bdad86231845cc4567bbeaedfcde722ed4\test_package.vcxproj]
3>test_package.obj : error LNK2019: unresolved external symbol glfwCreateWindow referenced in function main [C:\Us
ers\user\Source\conan-glfw\test_package\build\99c165bdad86231845cc4567bbeaedfcde722ed4\test_package.vcxproj]
3>test_package.obj : error LNK2019: unresolved external symbol glfwDestroyWindow referenced in function main [C:\U
sers\user\Source\conan-glfw\test_package\build\99c165bdad86231845cc4567bbeaedfcde722ed4\test_package.vcxproj]
3>C:\Users\user\Source\conan-glfw\test_package\build\99c165bdad86231845cc4567bbeaedfcde722ed4\bin\test_package.exe :
fatal error LNK1120: 4 unresolved externals [C:\Users\user\Source\conan-glfw\test_package\build\99c165bdad8623184
5cc4567bbeaedfcde722ed4\test_package.vcxproj]
```

I'm sure I'll figure it out in the end, but you have more experience building this way, so if you see any obvious mistakes you might save me some time 😄

Commit:
Update for a working multi-package

bschindler commented:
So I updated the recipe so that it actually works. Also, the way I reworked it it doesn't need the contained cmakelists.txt which in my opinion simplifies things a bit
Note, the include_pdb's option is not yet wired up so far. Will do this also at some point. I think it's more important to first decide on how/whether multi-packages get accepted or not. 

There was also some discussion about this topic on conan: 
https://github.com/conan-io/conan/issues/2328
https://github.com/conan-io/conan/issues/2327

ghost commented:
Now it works nicely and, looking at the documentation pull request mentioned in the second discussion you have listed, I suppose that build_type and runtime can be "officially" removed when creating this kind of recipe 😄
I like your solution, but we need to hear other bincrafters' thoughts.

bschindler commented:
For our usecase, we don't remove build_type officially as the package is a hybrid - single package where it needs to be and multi package where multi-package is needed. So it has to be conditionally removed which is what the recipe does. 

There is however some news on the multi-package front - I was not aware that the cmake integration offered by Visual Studio (VS 2017 onwards) changes how builds are dealt with. If a cmake project is opened in VS17 (not by creating a solution from cmake but with opening cmake directly), VS will create a full build directory per build type and runs cmake on each one of them. With that, single-build type packages can also be used in VS seamlessly. If this integration is robust enough, I see it as likely that we will switch to using that and ditch our multi-packages. So take this into account when checking with the wider bincrafters community.

ghost commented:
Oh, yes about build_type, that's what I meant. By saying "officially" I was just referring to the removal being documented in the pull request.
Sure, if it turns out that this can be handled in VS I think the advantage will be simpler recipes 👍

solvingj commented:
Glad to hear the Cmake integration might solve it for you.  Here's an example of using the VS Cmake integration with Conan.  It requires a the use of `cmake-conan` .  If you have more questions about this, please do open an issue in the community issues repo.
